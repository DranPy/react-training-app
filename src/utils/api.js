import axios from 'axios';
// NOTE: add a symlink
// import { host, port } from '../json-server/db.json';

//const URL = `http://${host}:${port}`;
const baseURL = `http://localhost:4000`;

export default axios.create({ baseURL });
