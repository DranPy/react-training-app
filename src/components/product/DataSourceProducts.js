import { Component } from 'react';
import api from '../../utils/api-helper';

class DataSourceProducts extends Component {
  static defaultProps = {
    isFetchImmediately: false,
  };

  state = {
    isLoading: false,
    products: [],
  };

  fetchProducts = async () => {
    this.setState({ isLoading: true });
    const products = await api.getAllProducts();
    this.setState({
      isLoading: false,
      products,
    });

    return products;
  };

  componentDidMount() {
    const { isFetchImmediately } = this.props;
    if (isFetchImmediately) {
      this.fetchProducts();
    }
  }

  render() {
    const { isLoading, products } = this.state;

    const newProps = {
      fetchProducts: this.fetchProducts,
      isLoading,
      products,
    };

    return this.props.children(newProps);
  }
}

export default DataSourceProducts;
