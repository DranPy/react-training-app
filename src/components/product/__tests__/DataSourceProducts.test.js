import { shallow, mount } from 'enzyme';
import React, { Component } from 'react';

import DataSourceProducts from '../DataSourceProducts';

import api from '../../../utils/api-helper';

jest.mock('../../../utils/api-helper');

describe('<DataSourceProducts />', () => {
  let WrapperComponent;

  class MockComponent extends Component {
    render() {
      return <div>Component</div>;
    }
  }

  it('should correct render', () => {
    const wrapper = mount(
      <DataSourceProducts>{data => <MockComponent {...data} />}</DataSourceProducts>
    );
    expect(wrapper).toMatchSnapshot();
  });

  it('should fetch products immediately', () => {
    const wrapper = mount(
      <DataSourceProducts isFetchImmediately={true}>
        {data => <MockComponent {...data} />}
      </DataSourceProducts>,
      { disableLifecycleMethods: true }
    );
    const instance = wrapper.instance();
    jest.spyOn(instance, 'fetchProducts');
    instance.componentDidMount();
    expect(instance.fetchProducts).toHaveBeenCalledTimes(1);
  });

  it('should not fetch products immediately', () => {
    const wrapper = mount(
      <DataSourceProducts isFetchImmediately={false}>
        {data => <MockComponent {...data} />}
      </DataSourceProducts>,
      { disableLifecycleMethods: true }
    );
    const instance = wrapper.instance();
    jest.spyOn(instance, 'fetchProducts');
    instance.componentDidMount();
    expect(instance.fetchProducts).not.toHaveBeenCalledTimes(1);
  });

  it('should fetch products to state', async () => {
    const mockProducts = [{ name: 'test 1' }, { name: 'test 2' }];
    api.getAllProducts.mockReturnValue(Promise.resolve(mockProducts));
    const wrapper = mount(
      <DataSourceProducts>{data => <MockComponent {...data} />}</DataSourceProducts>
    );
    const instance = wrapper.instance();
    await instance.fetchProducts();
    expect(wrapper.state('products')).toEqual(mockProducts);
  });
});
