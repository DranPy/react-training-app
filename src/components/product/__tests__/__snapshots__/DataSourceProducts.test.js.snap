// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<DataSourceProducts /> should correct render 1`] = `
<DataSourceProducts
  isFetchImmediately={false}
>
  <MockComponent
    fetchProducts={[Function]}
    isLoading={false}
    products={Array []}
  >
    <div>
      Component
    </div>
  </MockComponent>
</DataSourceProducts>
`;
